### PR TITLE
expanded package development docs

### PR DESCRIPTION
**Pull request by milktrader**
- makes mention earlier on the importance of a `REQUIRE` file and the preference of `v0.3` over `v0.3-`.
- combines tagging and publishing to emphasize importance of tagging.
- some more package naming guidelines for creative names.

To keep track of editorial input, here is a list:
- [x] remove _Creative names are cool_ section (or maybe reword it?) or not
- [ ] ~~enforce 92 char limit per line~~

### CODE REVIEW

**Comment by milktrader** [comment]
https://github.com/JuliaLang/METADATA.jl/issues/2481

File: doc/manual/packages.rst [hunk 1]
```diff
@@ -574,7 +574,18 @@ going to have to do a *force push*:
...
 ----------------------
 
-Guidelines for Naming a Package
+REQUIRE speaks for itself
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You should have a ``REQUIRE`` file in your package repo, with a bare minimum directive of what Julia version to include.
+As Julia moves to different versions, many packages include incorrect version requirements in their ``REQUIRE`` file. The most
+common mistake is to include pre-release versions of Julia after a release has been officially completed. If Julia has 
+officially released version ``0.3``, for example, then all the Julia-specific requirements in the ``REQUIRE`` file should not
```
Comment (tkelman) [inline doc/manual/packages.rst]: Should we replace these with generic ``0.x`` and ``0.y`` so we don't have to keep updating them?

File: doc/manual/packages.rst [hunk 2]
```diff
@@ -621,11 +632,22 @@ not domain experts*.
...
 
+6. Creative names are cool, but confusion isn't so much.
+
+  * Many modern programming languages such as Ruby and Python have popular packages
+    that have entertaining names. Python's ``pickle`` package isn't about pickles, it's 
+    for serializing (or pickling) object structures. Ruby's ``Sinatra`` package is named 
+    after the popular American singer and is a popular web framework that competes with Ruby's 
+    ``Rails`` web framework (another creative name).
+
+   * If being a creative package namer isn't your style, it's okay to not force it.
```
Comment (tkelman) [inline doc/manual/packages.rst]: does rst care that this is one space further indented than the previous bullet?

**Comment by tkelman** [comment]
This is excellent, thank you for doing this.

**Comment by ScottPJones** [comment]
:+1: Very useful information, esp. since I've been asked to move something to a Package instead of simply removing it...

**Comment by tkelman** [comment]
ah and double-check that this passes `make check-whitespace` locally before pushing

File: doc/manual/packages.rst (hunk 1 above)
Comment (milktrader) [inline doc/manual/packages.rst]: Yep, good point. I felt queasy using 3 and 4.

File: doc/manual/packages.rst (hunk 2 above)
Comment (milktrader) [inline doc/manual/packages.rst]: I'll fix that, thanks. Didn't catch it earlier.

**Comment by milktrader** [comment]
Yeah, I see I have an unseemly amount of trailing whitespace, which I think I got most of it visually in vi. How does one use `make check-whitespace`?

**Comment by milktrader** [comment]
Do I do this?
`git rebase --whitespace=fix package_docs`

**Comment by milktrader** [comment]
Oh, forget it. I was running `make check-whitespace` in the wrong directory,

This was the brainchild of @jakebolewski iirc.

**Comment by milktrader** [comment]
btw, the trailing whitespace check is pretty cool.

File: doc/manual/packages.rst [hunk 3]
```diff
@@ -570,11 +570,21 @@ going to have to do a *force push*:
...
 ----------------------
 
-Guidelines for Naming a Package
+REQUIRE speaks for itself
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You should have a ``REQUIRE`` file in your package repo, with a bare minimum directive of what Julia version to include.
+As Julia moves to different versions, many packages include incorrect version requirements in their ``REQUIRE`` file. The most
+common mistake is to include pre-release versions of Julia after a release has been officially completed. If Julia has
+officially released version ``0.x``, for example, then all the Julia-specific requirements in the ``REQUIRE`` file should not
```
Comment (IainNZ) [inline doc/manual/packages.rst]: Maybe reword this more along the lines of, `if Julia version 0.x has been released then you should use 0.x, not 0.x-. 0.x- indicates that you package is compatible with the yet....` and then the rest. Just so its more what you should do, rather than not do, is my thinking here.

**Comment by IainNZ** [comment]
Good stuff! You coming to JuliaCon again this year?

File: doc/manual/packages.rst (hunk 3 above)
Comment (milktrader) [inline doc/manual/packages.rst]: Got it, will do.

**Comment by milktrader** [comment]
yep, but gotta leave on Friday to make my Century ride on Saturday :smile:

**Comment by johnmyleswhite** [comment]
Nice work, @milktrader.

**Comment by milktrader** [comment]
Thanks John. 

I forgot to add `[av skip]` so this may take a while to green up. I did run `make check-whitespace` locally so travis should be fine.

File: doc/manual/packages.rst [hunk 4]
```diff
@@ -570,11 +570,19 @@ going to have to do a *force push*:
...
-
 Creating a new Package
 ----------------------
 
-Guidelines for Naming a Package
+REQUIRE speaks for itself
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You should have a ``REQUIRE`` file in your package repo, with a bare minimum directive of what Julia version to include.
+When Julia version ``0.x`` is released, use ``julia 0.x`` in your ``REQUIRE``, not ``julia 0.x-``. You may consider using
```
Comment (nalimilan) [inline doc/manual/packages.rst]: I would explain what `0.x-` means before saying not to use it. In this version, my reaction was: "why do you tell me about this if I'm not supposed to use it." :-) So, first say what to do to depend on a released version, and then what to (possibly) do if you depend on a development version.

File: doc/manual/packages.rst [hunk 5]
```diff
@@ -621,11 +629,22 @@ not domain experts*.
...
 
+6. Creative names are cool, but confusion isn't so much.
+
+  * Many modern programming languages such as Ruby and Python have popular packages
+    that have entertaining names. Python's ``pickle`` package isn't about pickles, it's
+    for serializing (or pickling) object structures. Ruby's ``Sinatra`` package is named
+    after the popular American singer and is a popular web framework that competes with Ruby's
+    ``Rails`` web framework (another creative name).
+
+  * If being a creative package namer isn't your style, it's okay to not force it.
```
Comment (nalimilan) [inline doc/manual/packages.rst]: This sounds like you should find a funny name, and only use a boring one if you're not smart enough. Also the first paragraph isn't clear about whether you consider this a good idea or not. I'd prefer people to use descriptive names where possible instead of what you explain about `pickle` and Sinatra`, and AFAICT this was the logic of the existing text. how about saying "use a descriptive package name instead of a creative one where applicable to help users finding out about it".

File: doc/manual/packages.rst [hunk 6]
```diff
@@ -621,11 +629,22 @@ not domain experts*.
...
+
 Generating the package
 ~~~~~~~~~~~~~~~~~~~~~~
 
-Suppose you want to create a new Julia package called ``FooBar``.
-To get started, do :func:`Pkg.generate(pkg,license) <Pkg.generate>` where ``pkg`` is the new package name and ``license`` is the name of a license that the package generator knows about::
+Suppose you want to create a new Julia package called ``FooBar``.  To get started, do
```
Comment (nalimilan) [inline doc/manual/packages.rst]: You added line breaks, but the lines are still too wide according to the 92 chars limit (same below).

File: doc/manual/packages.rst (hunk 4 above)
Comment (milktrader) [inline doc/manual/packages.rst]: Ok, thanks. I'll clean this up.

File: doc/manual/packages.rst (hunk 5 above)
Comment (milktrader) [inline doc/manual/packages.rst]: That's funny, I see your point. 

Creative names are for the cool kids, but if it's not your thing, it's ok. Just be descriptive and boring. (sort of paraphrasing you here).

I'm happy to remove this entire section. I thought I'd throw it in at the end of package naming guidelines to open the door to being a bit more creative. Oftentimes being descriptive devolves into an acronym party and we end up with packages like ABC.jl DEF.jl GHIK.jl and so forth. 

Given the option, I'd prefer a clever metaphor-like name to an acronym. I think they are equally less-descriptive. But when I see an acronym I cringe at having to decipher it but when I see a suggestive metaphor, I'm happy and intrigued. I _want_ to check out `pickle` and `cucumber` just because they sound cool. 

If you were to invent a javascript compiler and called it `JavaComp` it would be somewhat descriptive. And admittedly boring. But if you called it `CoffeeScript` then you will have achieved some marketing panache. 

Anyway, I'm beginning to ramble. I'm happy to keep editing away at this section and squash commits once it achieves some consensus!

File: doc/manual/packages.rst (hunk 6 above)
Comment (milktrader) [inline doc/manual/packages.rst]: Yeah, I broke up some other text with line breaks because it was becoming hard to look at. I'm unfamiliar with the 92 chars limit though.

Comment (milktrader) [inline doc/manual/packages.rst]: I added this in a todo list at the top of this PR.

File: doc/manual/packages.rst (hunk 5 above)
Comment (nalimilan) [inline doc/manual/packages.rst]: I'm not opposed to creative names when no obvious descriptive name comes to mind, but I'd simply recommend using explicit names where possible, that's all. For example, I'm glad `DataFrames.jl` is called that way rather than `Pandas.jl`.

(I don't think the marketing plays a big role when there are several thousands packages like in R's CRAN. At some point there is too much information for you to care...)

Comment (ScottPJones) [inline doc/manual/packages.rst]: Well, `CoffeeScript` *is* descriptive, as are variants with `Mocha`, `Latte`, or whatever your favorite is... one doesn't *have* to be boring just to be descriptive.  `pickle`, `cucumber`, and `Pandas` sound cute, but also are a royal pain if you are trying to Google search for the computer term!
Maybe your section could elaborate a bit on that, providing descriptive, but not necessarily boring names (but saying that boring is fine also, as long as it is descriptive).

Comment (milktrader) [inline doc/manual/packages.rst]: So a title along the lines of ...

>__Creative names when descriptive ones are elusive__
> __Consider an apt metaphor when being descriptive names become long, etc

Comment (milktrader) [inline doc/manual/packages.rst]: Here's a result for searching on `pickle`
screen shot 2015-06-16 at 11 34 33 am

Comment (ScottPJones) [inline doc/manual/packages.rst]: Sounds good.

Comment (ScottPJones) [inline doc/manual/packages.rst]: You might also caution against naming something using a common term... (Like `Mumps` for the name of a language, used by the medical community... funny, haha, but horrible when you want to search [although on-line search didn't exist when it was named])

Comment (ScottPJones) [inline doc/manual/packages.rst]: Right, but not all pages about "pickle" the Python serialization protocol necessarily have the word "Python" in them... that's the problem... not that you can't find *some* pages, or even most, by adding extra terms.

**Comment by milktrader** [comment]
I'm starting to think that it would be better to remove the creative naming from this PR and just focus on tagging and julia version syntax. I could open a separate PR for expanding the package naming docs.

**Comment by ScottPJones** [comment]
Sure, no reason to hold up this good work for naming

**Comment by milktrader** [comment]
All squashed up and waiting for travis to check for trailing whitespace.

File: doc/manual/packages.rst [hunk 7]
```diff
@@ -570,11 +570,21 @@ going to have to do a *force push*:
...
-
 Creating a new Package
 ----------------------
 
-Guidelines for Naming a Package
+REQUIRE speaks for itself
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You should have a ``REQUIRE`` file in your package repo, with a bare minimum directive of what Julia version to include.
+When Julia version ``0.x`` is released, use ``julia 0.x`` in your ``REQUIRE``. You may consider using ``julia 0.y-`` at this
```
Comment (tkelman) [inline doc/manual/packages.rst]: No, the only reason to add a requirement on a not-yet-released version is if your package is actually incompatible with the existing releases. I think much of the confusion stems from an intuition-based reading of `0.3-` as "0.3 or newer," without realizing that the first item is a lower bound by default, and upper bounds are a separate second item if they're listed at all.

Comment (milktrader) [inline doc/manual/packages.rst]: Ok, I'll grok this and fix it. Thanks.

Comment (milktrader) [inline doc/manual/packages.rst]: I was letting the `x` and `y` confuse me. Okay, I get it. By saying `0.y-` you are _not_ supporting `0.x`.

**Comment by milktrader** [comment]
https://github.com/JuliaLang/julia/issues/11734

**Comment by milktrader** [comment]
I'm going to pare down the entire `REQUIRE` section so we can move on.

File: doc/manual/packages.rst [hunk 8]
```diff
@@ -570,11 +570,36 @@ going to have to do a *force push*:
...
 ----------------------
 
-Guidelines for Naming a Package
+REQUIRE speaks for itself
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You should have a ``REQUIRE`` file in your package repo, with a bare minimum directive of what Julia
+version you expect your users to be running for the package to work. When Julia version ``0.x`` is released,
+use ``julia 0.x`` in your ``REQUIRE`` file. Typically, once a stable release is officially released,
+developers abaondon support for previous versions in future releases of their packages.
```
Comment (tkelman) [inline doc/manual/packages.rst]: abandon

Comment (tkelman) [inline doc/manual/packages.rst]: Maybe a more delicate way of saying this would be "Package developers should set, as a floor, the lowest version of Julia they regularly test their package against. Travis-CI (link) enables automatic testing against multiple versions of Julia, and `Pkg.generate` pre-populates a configuration file for it."

Comment (tkelman) [inline doc/manual/packages.rst]: Maybe just delete this last sentence since it's pretty well-covered by the next few paragraphs. We have a link to Travis below, but would be good to elaborate that it allows automatic testing against multiple Julia versions.

**Comment by milktrader** [comment]
Yeah, abandon is not a good word.

**Comment by milktrader** [comment]
I think remove the last sentence and maybe provide some rationale for why put it in at all.

**Comment by milktrader** [comment]
https://github.com/JuliaLang/METADATA.jl/pull/2744#issuecomment-113356425

**Comment by milktrader** [comment]
s/Typically, once a stable release is officially released,
+developers abaondon support for previous versions in future releases of their packages./
If you leave this information out, <something happens that isn't ideal>

**Comment by milktrader** [comment]
or, 
>Typically, once a stable release is officially released, developers begin testing their code on that >version. In the case where you don't provide any information about a julia version, something >happens that isn't ideal

**Comment by ScottPJones** [comment]
But, what happens that isn't ideal exactly (sounds rather coy!)

**Comment by milktrader** [comment]
https://github.com/rened/METADATA.jl/commit/d5ad2def94362442e1b785a893b622030476d130

**Comment by milktrader** [comment]
That's not the intended text. I'm not sure what happens. It's a,b,c or d in the EDIT above.

**Comment by milktrader** [comment]
https://github.com/JuliaLang/METADATA.jl/pull/2748

**Comment by tkelman** [comment]
Travis defaults to testing against release at the moment. We still need to add in version-number-based testing as an alternative to make it possible to still test against 0.3 even after 0.4.0 gets released.

Adding a `julia 0.x` line to your REQUIRE file isn't really directly related to Travis. It's a statement that "This is the minimum version of Julia for which this package is developed (and tested)." I'm not opposed to packages still attempting to support 0.2 if the developer wants to do so, but that would require regular testing which currently isn't so easy via Travis, you have to do it manually.

**Comment by milktrader** [comment]
ok, so the purpose is informational, and apparently consequential in the case when `Pkg.update()` is run.

File: doc/manual/packages.rst [hunk 9]
```diff
@@ -814,18 +806,19 @@ It also creates a new version entry in your local ``METADATA`` repo for ``FooBar
...
 
-If there is a ``REQUIRE`` file in your package repo, it will be copied into the appropriate spot in ``METADATA`` when you tag a version.
-Package developers should make sure that the ``REQUIRE`` file in their package correctly reflects the requirements of their package, which will automatically flow into the official metadata if you're using :func:`Pkg.tag`.
-See the `Requirements Specification <#man-package-requirements>`_ for the full format of ``REQUIRE``.
+The :func:`Pkg.tag` command takes an optional second argument that is either an explicit version number object
+like ``v"0.0.1"`` or one of the symbols ``:patch``, ``:minor`` or ``:major``.  These increment the patch, minor
+or major version number of your package intelligently.
+
+Adding a tagged version of your package will expedite the official registration into METADATA.jl by collaborators. It is
+strongly recommended that you complete this process, regardless if your package is completely ready for an official release.
```
Comment (tkelman) [inline doc/manual/packages.rst]: Should we add a link to semver.org for recommended best practices on version numbering?

Comment (tkelman) [inline doc/manual/packages.rst]: Actually nevermind maybe not, since the "Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable." bullet is easy to miss. Hopefully people can figure out when is best to do minor vs patch.

**Comment by milktrader** [comment]
This PR feels done for now. I've already taken out additions to the Package naming for a separate PR (not yet pulled, but marinating). So we have a re-org of tagging to be up front with publishing now and a brief (and I think - correct me if wrong ) and accurate overview of including Julia versions in the REQUIRE.

**Comment by tkelman** [comment]
Agreed, +1 for merging now. I'll wait a little while for any other feedback before doing so, if no one beats me to it.

File: doc/manual/packages.rst [hunk 10]
```diff
@@ -570,11 +570,38 @@ going to have to do a *force push*:
...
 
-
 Creating a new Package
 ----------------------
 
-Guidelines for Naming a Package
+REQUIRE speaks for itself
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You should have a ``REQUIRE`` file in your package repo, with a bare minimum directive of what Julia
```
Comment (IainNZ) [inline doc/manual/packages.rst]: A thought, take or leave it: maybe spell  `repository` out instead of `repo`.

Comment (milktrader) [inline doc/manual/packages.rst]: Yes.

**Comment by milktrader** [comment]
Travis should pass, then I'll squash.

**Comment by milktrader** [comment]
squash in progress ...

````bash
[julia (package_docs)] 
✈  git checkout master
✈  git pull
✈  git checkout package_docs
✈  git rebase -i origin/master
# in this interactive mode, `s/commit/squash` all the lines _except_ the first one
✈  git push milktrader package_docs -f
````

**Comment by tkelman** [comment]
I think we're good to go here? Thanks so much for doing this!

**Comment by milktrader** [comment]
Thanks for all the feedback!
